### PR TITLE
Add ResultType protocol to facilitate protocol extension constraints

### PR DESCRIPTION
**Pull request by Thomvis**
Protocols are really powerful in Swift and with Swift 2.0 even more so. By adding the `ResultType` protocol, we can use the type in protocol extensions and in generic constraints.

The primary motivation for this PR is a refactoring of BrightFutures (which has Result as a dependency) that I am working on. With the help of the `ResultType` protocol, I can define a simple `Deferred<T>` type and add methods (such as `onSuccess` and `onFailure`) to it when `T: ResultType`. 
- Relevant code in BrightFutures: https://gist.github.com/Thomvis/0684c36159d173396297ub.com/Thomvis/BrightFutures/blob/swiftier-2.0/BrightFutures/Future.swift#L112
- Concise example:  https://gist.github.com/Thomvis/0684c36159d173396297

The contents of the protocol is debatable, but the current contents are the parts that I found helpful.

:star2: Let me know what you think!

### CODE REVIEW

File: Result/Result.swift [hunk 1]
```diff
@@ -131,6 +131,21 @@ public enum Result<T, Error: ErrorType>: CustomStringConvertible, CustomDebugStr
 	}
 }
 
+public protocol ResultType {
```
Comment (robrix) [inline Result/Result.swift]: Could you add the protocol to a new file, please?

File: Result/Result.swift [hunk 2]
```diff
@@ -131,6 +131,21 @@ public enum Result<T, Error: ErrorType>: CustomStringConvertible, CustomDebugStr
...
+	var value: Value? { get }
+	var error: Error? { get }
+	
+	init(value: Value)
+	init(error: Error)
+	
+	func analysis<U>(@noescape ifSuccess ifSuccess: Value -> U, @noescape ifFailure: Error -> U) -> U
+}
+
+extension Result: ResultType { }
```
Comment (robrix) [inline Result/Result.swift]: Could you please declare this on `Result` rather than in an extension?

File: Result/Result.swift (hunk 1 above)
Comment (robrix) [inline Result/Result.swift]: Also, simple documentation comments for the protocol and its members would be greatly appreciated :bow:

**Comment by robrix** [comment]
Thanks for the PR, @Thomvis! @antitypical/result, I’d love to know what you think. I’m quite :+1: on what this enables for APIs using `Result`.

**Comment by mdiep** [comment]
I'm sad that a protocol is required to do this, but very :+1: on what it enables.

**Comment by Thomvis** [comment]
Thanks for the feedback. I think they're all valid and will follow up over the weekend.

File: Result/Result.swift [hunk 3]
```diff
@@ -131,6 +131,21 @@ public enum Result<T, Error: ErrorType>: CustomStringConvertible, CustomDebugStr
 	}
 }
 
+public protocol ResultType {
+	typealias Value
+	typealias Error: ErrorType
+	
+	var value: Value? { get }
+	var error: Error? { get }
```
Comment (robrix) [inline Result/Result.swift]: Given `analysis`, you can provide `value` and `error` in a protocol extension. That reduces the requirements to two introduction rules (the `init`s) and one elimination rule (`analysis`) covering both alternatives, which is its logical minimum :tada:

**Comment by Thomvis** [comment]
I've updated the PR according to your feedback. I'm not very confident about the documentation, so feel free to give feedback or change it if you decide to merge it in.

**Comment by robrix** [comment]
Needs a merge from `master` :pray:

File: Result/ResultType.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
+//  Copyright (c) 2015 Rob Rix. All rights reserved.
+
+/// A type that can represent either failure with an error or success with a result value.
+public protocol ResultType {
+	typealias Value
+	typealias Error: ErrorType
+	
+	/// Constructs a succesful result wrapping a `value`.
```
Comment (robrix) [inline Result/ResultType.swift]: Nitpick: `successful`

**Comment by robrix** [comment]
Docs are good, modulo the one typo; thanks so much @Thomvis!

**Comment by Thomvis** [comment]
:white_check_mark: corrected typo
:white_check_mark: merged from `master`

**Comment by neilpa** [comment]
:sparkling_heart: the direction here

**Comment by robrix** [comment]
It is a delight to press this merge button! :tada:

**Pull request merged**